### DISCUSSION TITLE
I want copilot to automatically find code blocks that can be optimized/simplified.

### DISCUSSION THREAD

**Discussion by wangmingliang-ms** [community/community#109538] [Copilot Conversations]
### Select Topic Area

Show & Tell

### Body

I want copilot to automatically help me find code blocks that can be optimized/simplified for better performance or readability, or because of newly added features. 

I've tried it a bit and it works pretty well, e.g.
It can tell me that I can actually use `record` introduced in `Java 14` to simplify `POJO`s. It can also tell me about the new `switch` introduced in Java 14 to simplify multiple `if-else`. 

**but I feel like it should be part of GitHub Copilot itself. Can you add this feature into GitHub Copilot?**

![image](https://github.com/community/community/assets/69189193/649bf732-c888-4985-ac9f-89c1607ef36e)

**Comment by dmm2321** [comment]
I have a long answer for you, but I think it will help. I'm new to this myself, but I read through a lot of official documentation, so I feel pretty confident with what I've learned. I hope it helps :)

So, you actually can optimize code using Copilot Chat. It's not "automatic" but only requires a quick command to do the task. There's a slash command '/optimize' that I just read about in a github blog article from last month. I think it's a recent addition because when you type '/' in the chat to list the possible commands, it's not listed. Though if you used Chat to ask something similar like "How can I optimize or improve my code", the results look to be similar, if not the same as using the /optimize command. The one word command is just simpler. 

Anyway, there are a couple of options. 
1. For a single block of code, you can highlight it, open the inline chat (Ctrl + I) and run the /optimize command. 
2. For an entire (single) file, if you open the Chat view panel (Ctrl + Alt + I) and again, type /optimize, it will review whatever file/tab is actively open in your window. If you have multiple files open, be sure to have the right tab open. 
3. To examine a program made up of multiple files, you can examine them all at once too, just use the VSCode's agent: @workspace in front of the /optimize command. (@workspace /optimize)

Below are the links I found the most informative. I suggest looking at the top 2 because they talk about how to get the best results from Copilot Chat (mainly a reminder of good programming standards so the code is easy to understand, with comments for the AI to easily follow your logic/intentions)

https://code.visualstudio.com/docs/copilot/prompt-crafting#_getting-the-most-out-of-copilot-inline-suggestions
https://code.visualstudio.com/docs/copilot/getting-started-chat
https://github.blog/changelog/2024-01-30-code-faster-and-better-with-github-copilots-new-features-in-visual-studio/
https://code.visualstudio.com/docs/editor/workspaces

**Reply by wangmingliang-ms** [reply]
thanks, @dmm2321. **the key point "automatic".** The existing copilot operates in a rather passive manner. It requires me to actively prompt it before it provides any information. 
I will only call that command when I know it can be optimized. But **If I wasn't aware that `Java 14` introduced the new `record`, I would never call that command (`/optimize`), and that code would never be optimized.**

This feature is particularly beneficial for those who are just starting to learn the language.

**Reply by misrasaurabh1** [reply]
I am late to the discussion but I am building the above automated code optimizer for Python - [https://codeflash.ai](https://codeflash.ai)

**Comment by jfullstackdev** [comment]
while the intention is good, and it works with personal projects, it would, I guess be more devastating in an already established code, so rather than having that quick fix, it would be more of like prompting, particularly nowadays the AI, in Agent mode, like I always use GPT-4.1,

this way, the AI, when ask to learn a certain feature, say in that Module, is context aware and when you ask it like, is there any better refactor, it can make it without breaking or changing the existing functionality, in simple projects, this might be overkill but in actual Dev, this is the standard, we can't even simply change code or refactor in the quickest way then merge it in Prod, because that's too dangerous 

other than that, what you might be referring are the static analyzers like linters or SonarQube that can analyze code based on rules, so this, together with AI, produce more robust code
